### PR TITLE
WT-13616 Make page->read_gen private to the eviction module

### PR DESCRIPTION
**Pull request by Siddhartha8899**
Summarize the reason behind this change (this might be the problem you're solving, or the context around the request) and the solution you have chosen.

### CODE REVIEW

File: src/evict/evict_inline.h [hunk 1]
```diff
@@ -76,20 +93,19 @@ __wt_evict_read_gen_new(WT_SESSION_IMPL *session, WT_PAGE *page)
...
- *     Indicate if the cache is stuck (i.e., eviction not making progress).
+ * __wt_readgen_evict_soon --
+ *     Return whether a page's read generation makes it eligible for immediate eviction. Read
+ *     generations reserve a range of low numbers for special meanings and currently - with the
+ *     exception of the generation not being set - these indicate the page may be evicted
+ *     immediately.
  */
 static WT_INLINE bool
-__wt_evict_cache_stuck(WT_SESSION_IMPL *session)
+__wt_readgen_evict_soon(uint64_t *readgen)
```
Comment (Siddhartha8899) [inline src/evict/evict_inline.h]: It is removed from `btree_inline.h` in a later ([11ca8bb](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/11ca8bb682558dfbbcc188aa55d6f4e49986ef2a))commit. Sorry about that (this should have been the first commit)

**Review by keitharnoldsmith — COMMENTED** [review]
Thanks Sid. Please make sure to fill out the testing/risk template for the PR.

Suggestion (it could be a separate ticket): I think the code would be easier to read if we used `WT_READGEN_URGENT` (`WT_READGEN_SOON`) instead of `WT_READGEN_OLDEST`.  That way it's more obvious what we use that value for.

Maybe there are places where we actually want the oldest possible readgen value. So I would just `#define WT_READGEN_URGENT WT_READGEN_OLDEST`.

File: src/evict/evict_lru.c [hunk 2]
```diff
@@ -2145,11 +2145,11 @@ __evict_try_queue_page(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, WT_REF *
      * somehow leave a page without a read generation.
      */
     if (__wt_atomic_load64(&page->read_gen) == WT_READGEN_NOTSET)
-        __wt_evict_read_gen_new(session, page);
+        __wti_evict_read_gen_new(session, page);
 
     /* Pages being forcibly evicted go on the urgent queue. */
     if (modified &&
-      (__wt_atomic_load64(&page->read_gen) == WT_READGEN_OLDEST ||
+      (__wti_evict_page_soon_flagged(&page->read_gen) ||
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/evict/evict_lru.c:2152]: Thanks @keitharnoldsmith, I agree, I have now mentioned it [here](https://github.com/wiredtiger/wiredtiger/pull/11087#issuecomment-2398964139) under the `What, if anything, are you concerned about that you'd like the reviewer to focus on?
` section.
 I had a tough time dealing with the different read generations, since this is moving towards a functional change. I am happy to revert the [1fc5939](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/1fc5939dc3d87cec047b84688ea46854f36cee3a) commit for now and only keep the refactoring changes here. This means we would have a getter of read_gen outside the eviction code [here](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/1fc5939dc3d87cec047b84688ea46854f36cee3a#diff-9691a0692cb4c26132c4c27aa260cc6441c2ace4b04487d51769fe84869c624aR172). Let me know you thoughts.

File: src/evict/evict_inline.h [hunk 3]
```diff
@@ -65,31 +72,40 @@ __wt_evict_read_gen_bump(WT_SESSION_IMPL *session, WT_PAGE *page)
...
-    uint32_t tmp_evict_aggressive_score;
+    uint64_t gen;
 
-    evict = S2C(session)->evict;
-    tmp_evict_aggressive_score = __wt_atomic_load32(&evict->evict_aggressive_score);
-    WT_ASSERT(session, tmp_evict_aggressive_score <= WT_EVICT_SCORE_MAX);
-    return (
-      tmp_evict_aggressive_score == WT_EVICT_SCORE_MAX && F_ISSET(evict, WT_EVICT_CACHE_HARD));
+    WT_READ_ONCE(gen, *readgen);
+    return (gen != WT_READGEN_NOTSET && gen < WT_READGEN_START_VALUE);
```
Comment (ajmorton) [inline src/evict/evict_inline.h:98]: I'll defer to you on the final decision. You have a lot more context on this change.

I'm wondering if we could define API functions
```
__wt_evict_readgen_is_init()
__wt_evict_readgen_is_evict_soon()
__wt_evict_readgen_is_wont_need()
```
or a generic
```
__wt_readgen_is(readgen, WT_READGEN_WONT_NEED)
```
here.

My reasoning is I don't see many new special `WT_READGEN_` values being added in future, and if they are perhaps we add a `WT_MUST_KEEP` value which would break the `< WT_READGEN_START_VALUE` check. 
This would let us be more specific in what we're checking. The line above for example would become
`return (__wt_evict_readgen_is(gen, WT_READGEN_OLDEST) || __wt_evict_readgen_is(gen, WT_READGEN_WONT_NEED));`

File: src/include/btmem.h [hunk 4]
```diff
@@ -13,8 +13,8 @@
 /* AUTOMATIC FLAG VALUE GENERATION START 0 */
 #define WT_READ_CACHE 0x0001u
 #define WT_READ_IGNORE_CACHE_SIZE 0x0002u
-#define WT_READ_NOTFOUND_OK 0x0004u
-#define WT_READ_NO_GEN 0x0008u
+#define WT_READ_INTERNAL_OP 0x0004u
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/include/btmem.h]: Updated in [79bded7](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/79bded74b464cbbb3732090e19ac908103ebcae1)

File: src/btree/bt_read.c [hunk 5]
```diff
@@ -503,27 +504,14 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
                  * If the page was read by this retrieval or was pulled into the cache via the
                  * pre-fetch mechanism, count that as a page read directly from disk.
                  */
-                if (F_ISSET_ATOMIC_16(page, WT_PAGE_PREFETCH) ||
-                  __wt_atomic_load64(&page->read_gen) == WT_READGEN_NOTSET)
+                if (F_ISSET_ATOMIC_16(page, WT_PAGE_PREFETCH) || new_read)
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_read.c]: Updated in [79bded7](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/79bded74b464cbbb3732090e19ac908103ebcae1)

File: src/btree/bt_read.c [hunk 6]
```diff
@@ -503,27 +504,14 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
...
-             */
-            if (__wt_atomic_load64(&page->read_gen) == WT_READGEN_NOTSET) {
-                if (wont_need)
-                    __wt_atomic_store64(&page->read_gen, WT_READGEN_WONT_NEED);
-                else
-                    __wt_evict_read_gen_new(session, page);
-            } else if (!LF_ISSET(WT_READ_NO_GEN))
-                __wt_evict_read_gen_bump(session, page);
+
+            /* Tell eviction we're using the page so it can update the page state. */
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_read.c]: Updated in [79bded7](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/79bded74b464cbbb3732090e19ac908103ebcae1)

File: src/evict/evict_inline.h [hunk 7]
```diff
@@ -104,6 +120,75 @@ __wt_evict_page_soon(WT_SESSION_IMPL *session, WT_REF *ref)
...
+            __wt_atomic_store64(&page->read_gen, WT_READGEN_WONT_NEED);
+        else
+            __wti_evict_read_gen_new(session, page);
+    } else if (!internal_only)
+        __wti_evict_read_gen_bump(session, page);
+}
+
+/*
+ * __wt_evict_page_init --
+ *     Initialize eviction state for a newly created page.
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/evict/evict_inline.h:163]: Updated

File: src/evict/evict_inline.h [hunk 8]
```diff
@@ -65,31 +72,40 @@ __wt_evict_read_gen_bump(WT_SESSION_IMPL *session, WT_PAGE *page)
...
- *     Indicate if the cache is stuck (i.e., eviction not making progress).
+ * __wti_evict_page_soon_flagged --
+ *     Return whether a read generation value makes a page eligible for immediate eviction. Read
+ *     generations reserve a range of low numbers for special meanings and currently - with the
+ *     exception of the generation not being set - these indicate the page may be evicted
+ *     immediately.
  */
 static WT_INLINE bool
-__wt_evict_cache_stuck(WT_SESSION_IMPL *session)
+__wti_evict_page_soon_flagged(uint64_t *readgen)
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/evict/evict_inline.h]: Updated in [79bded7](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/79bded74b464cbbb3732090e19ac908103ebcae1)

File: src/btree/bt_split.c [hunk 9]
```diff
@@ -1433,13 +1433,12 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
...
-     * was a forced eviction, in which case we leave the new page with the read generation unset.
-     * Eviction will set the read generation next time it visits this page.
+     * The re-instantiated page inherits the eviction state (LRU position) from the original page,
+     * unless this was a forced eviction, in which case we leave the new page with the default
+     * initialization.
      */
-    WT_READ_ONCE(orig_read_gen, orig->read_gen);
-    if (!__wt_readgen_evict_soon(&orig_read_gen))
-        __wt_atomic_store64(&page->read_gen, orig_read_gen);
+    if (!__wt_evict_page_soon_flagged(orig))
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_split.c]: Updated in [acfdc42](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/acfdc42905a9aedaec442c3eb087cc4c205c68cf)

File: src/evict/evict_inline.h (hunk 3 above)
Comment (Siddhartha8899) [reply src/evict/evict_inline.h:98]: I have made some changes in [0012113](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/0012113bc68246830394367a1c264190c6bbdcb6) commit. Let me know what you think.

Since most individual checks for `READGEN_OLDEST` and `READGEN_WONT_NEED` are internal to eviction, with only one [instance](https://github.com/wiredtiger/wiredtiger/pull/11087/files#diff-9691a0692cb4c26132c4c27aa260cc6441c2ace4b04487d51769fe84869c624aR172) outside eviction checking just `READGEN_OLDEST`, I have made the folliwing changes - 
- Reverted [1fc5939](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/1fc5939dc3d87cec047b84688ea46854f36cee3a) changes.
- Renamed `WT_READGEN_OLDEST` to `WT_READGEN_EVICT_SOON`.
- The setter for `WT_READGEN_EVICT_SOON` remains (`__wt_evict_page_soon`), and I’ve added a getter: `__wt_evict_page_is_soon`.
- Renamed `__wt_evict_soon_flagged` to `__wt_evict_is_soon_or_wont_need`.
- I didn’t change the return condition `return (gen != WT_READGEN_NOTSET && gen < WT_READGEN_START_VALUE)` to check only `WT_READGEN_EVICT_SOON` and `WONT_NEED`, as I’m unsure if  read_gen could be below 100.

File: src/btree/bt_cursor.c [hunk 10]
```diff
@@ -169,8 +169,7 @@ __cursor_page_pinned(WT_CURSOR_BTREE *cbt, bool search_operation)
      * whether we correctly resolved the transaction becomes hard. It is easier to skip this check
      * in that instance.
      */
-    if (__wt_atomic_load64(&cbt->ref->page->read_gen) == WT_READGEN_OLDEST &&
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_cursor.c:172]: Renamed `WT_READGEN_OLDEST` to `WT_READGEN_EVICT_SOON` in [0012113](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/0012113bc68246830394367a1c264190c6bbdcb6)

File: src/evict/evict_inline.h (hunk 3 above)
Comment (Siddhartha8899) [reply src/evict/evict_inline.h:98]: I will update the APIs in the description once this is approved.

File: src/evict/evict_inline.h [hunk 11]
```diff
@@ -101,7 +127,72 @@ __wt_evict_page_soon(WT_SESSION_IMPL *session, WT_REF *ref)
...
+__wt_evict_page_init(WT_PAGE *page)
+{
+    __wt_atomic_store64(&page->read_gen, WT_READGEN_NOTSET);
+}
+
+/*
+ * __wt_evict_inherit_page_state --
+ *     When creating a new page from an existing page, for example during split, initialize the read
+ *     generation on the new page using the read generation of the original page, unless this was a
+ *     forced eviction, in which case we leave the new page with the default initialization.
```
Comment (ajmorton) [inline src/evict/evict_inline.h:185]:
```suggestion
 *     immediate eviction, in which case we leave the new page with the default initialization.
```
nit: WONT_NEED won't be put in the urgent queue so "forced" seems like too strong a word here

File: src/evict/evict_lru.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajmorton) [reply src/evict/evict_lru.c]: We've reverted 1fc5939 (discussion [here](https://github.com/wiredtiger/wiredtiger/pull/11087#discussion_r1798620656)). I've verified this behaviour is back to only queuing urgent pages and not `WONT_NEED` pages.

**Review by ajmorton — APPROVED** [review]
All my comments are about nits/grammar except for the one about soon/forced/urgent/immediate/oldest. Don't let that block you though. Happy to merge once the nitpicks are resolved.

File: src/btree/bt_page.c [hunk 12]
```diff
@@ -60,7 +60,7 @@ __wt_page_alloc(
     WT_RET(__wt_calloc(session, 1, size, &page));
 
     page->type = type;
-    __wt_atomic_store64(&page->read_gen, WT_READGEN_NOTSET);
+    __wt_evict_page_init(page);
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_page.c:63]: I do not want to expose the name `readgen` outside the eviction module.

File: src/btree/bt_cursor.c [hunk 13]
```diff
@@ -169,8 +169,7 @@ __cursor_page_pinned(WT_CURSOR_BTREE *cbt, bool search_operation)
...
-      !F_ISSET(session->txn, WT_TXN_PREPARE))
+    if (__wt_evict_page_is_soon(cbt->ref->page) && !F_ISSET(session->txn, WT_TXN_PREPARE))
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_cursor.c:172]: This is a good question. We have a lot of eviction terminology in the code that causes confusion. 

Here’s my take:
1. Urgent eviction refers to anything in the urgent queue. I wouldn’t want to rename OLDEST to URGENT since they represent different concepts.

2. During forced eviction - 
    - Place a page in the urgent queue, making it part of **urgent eviction**, or
    - Directly evict the page, making it a **forced eviction**.
    - Forced eviction can only occur if either `READGEN_OLDEST` or `READGEN_WONT_NEED` is set on the page.

Therefore, the logic is:
forced eviction → (`OLDEST` | `WONT_NEED`).

We have renamed `OLDEST` to `EVICT_SOON`, as that better describes its purpose. In terms of priority, "evict soon" feels more immediate than "won't need".

Basically - 

- EVICT_SOON | WONT_NEED → forced eviction.
- EVICT_SOON has an individual getter and setter, which is useful in the code.
- There is a WT_EVICT_CACHE_URGENT flag in forced eviction. Ideally, it should be renamed to WT_EVICT_CACHE_URGENT, but I won’t be addressing this in the current PR.

File: src/evict/evict_inline.h (hunk 11 above)
Comment (Siddhartha8899) [reply src/evict/evict_inline.h:185]: Explained above -
forced eviction -> (oldest | wont_need)

File: src/evict/evict_inline.h [hunk 14]
```diff
@@ -101,7 +127,72 @@ __wt_evict_page_soon(WT_SESSION_IMPL *session, WT_REF *ref)
...
+     */
+    if (__wt_atomic_load64(&page->read_gen) == WT_READGEN_WONT_NEED)
+        __wti_evict_read_gen_new(session, page);
+}
+
+/*
+ * __wt_evict_touch_page --
+ *     Tell eviction when we touch a page so it can update its eviction state for that page. The
+ *     caller may set flags indicating that it doesn't expect to need the page again or that it's an
+ *     internal read. The latter is used by operations such as compact, and eviction, itself, so
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/evict/evict_inline.h]: Updated in [c274b8a](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/c274b8a5c2080efd723edfd6a61a875e7d85545c)

File: src/evict/evict_inline.h [hunk 15]
```diff
@@ -20,34 +20,41 @@ __wt_evict_aggressive(WT_SESSION_IMPL *session)
 }
 
 /*
- * __evict_read_gen --
- *     Get the current read generation number.
+ * __wt_evict_cache_stuck --
+ *     Indicate if the cache is stuck (i.e., eviction not making progress).
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/evict/evict_inline.h]: Updated in [c274b8a](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/c274b8a5c2080efd723edfd6a61a875e7d85545c).
And you're getting attention-to-detail bonusly for being so thorough!

File: src/evict/evict_inline.h [hunk 16]
```diff
@@ -101,7 +127,72 @@ __wt_evict_page_soon(WT_SESSION_IMPL *session, WT_REF *ref)
...
-    __wt_atomic_store64(&ref->page->read_gen, WT_READGEN_OLDEST);
+    __wt_atomic_store64(&ref->page->read_gen, WT_READGEN_EVICT_SOON);
+}
+
+/*
+ * __wt_evict_page_first_dirty --
+ *     Tell eviction if this is the first time the page has been modified while in the cache. The
+ *     eviction mechanism will then update the page's eviction state as needed. In this function,
+ *     although the page was not initially required, it has now been modified, so we prefer to
+ *     retain it instead of evicting it immediately.
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/evict/evict_inline.h]: Updated in [c274b8a](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/c274b8a5c2080efd723edfd6a61a875e7d85545c)

File: src/evict/evict_inline.h [hunk 17]
```diff
@@ -101,7 +127,72 @@ __wt_evict_page_soon(WT_SESSION_IMPL *session, WT_REF *ref)
...
+ */
+static WT_INLINE void
+__wt_evict_page_first_dirty(WT_SESSION_IMPL *session, WT_PAGE *page)
+{
+    /*
+     * In the event we dirty a page which is flagged for eviction soon, we update its read
```
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/evict/evict_inline.h]: Explanation for terminology - https://github.com/wiredtiger/wiredtiger/pull/11087#discussion_r1798839954

File: src/btree/bt_cursor.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_cursor.c:172]: I have removed the word immediate as part of [5464ef4](https://github.com/wiredtiger/wiredtiger/pull/11087/commits/5464ef407b545435ba0c5e78833c4cbf01e356bc)

[the remark this answers is not in the document]
Comment (ajmorton) [reply src/btree/bt_cursor.c:172]: SGTM. You have a better handle on the evict code than I do :).
[These lines](https://github.com/wiredtiger/wiredtiger/pull/11087/files#diff-fba173dbad7997f847117319e1939715f1969181ab2937508a2c9303b201f017R2152) are throwing me off as I assumed `WT_READGEN_OLDEST`/`WT_READGEN_EVICT_SOON` meant "stick this page in the urgent queue ASAP". Perhaps we should also detect `WONT_NEED` pages here but that's outside the PR scope. Resolving the convo

[the remark this answers is not in the document]
Comment (Siddhartha8899) [reply src/btree/bt_cursor.c:172]: I see. I agree, we can handle this aspect later.
